app/gosns/gosns.go

```diff
@@ -485,6 +491,132 @@ func DeleteTopic(w http.ResponseWriter, req *http.Request) {
 
 }
 
+func PublishBatch(w http.ResponseWriter, req *http.Request) {
+	content := req.FormValue("ContentType")
+	topicArn := req.FormValue("TopicArn")
+
+	arnSegments := strings.Split(topicArn, ":")
+	topicName := arnSegments[len(arnSegments)-1]
+	topic, ok := app.SyncTopics.Topics[topicName]
+	if !ok {
+		createErrorResponse(w, req, "TopicNotFound")
+		return
+	}
+
+	batchMessageIdToMessageBody := make(map[string]string, 10)
+	batchMessageIdToMessageStructure := make(map[string]string, 10)
+	batchMessageIdToMessageAttributes := make(map[string]map[string]app.MessageAttributeValue, 10)
+	batchMessageIdToSubject := make(map[string]string, 10)
+
+	for memberIndex := 1; true; memberIndex++ {
+		if memberIndex > 10 {
+			createErrorResponse(w, req, "TooManyEntriesInBatchRequest")
+			return
+		}
+		thisMessageFormKey := "PublishBatchRequestEntries.member." + strconv.Itoa(memberIndex)
+
+		batchMessageId := req.FormValue(thisMessageFormKey + ".Id")
+		if batchMessageId == "" {
+			// This is a required field, its absence likely indicates there are no further entries.
+			// TODO: how to tell the difference in case this is validation failure?
+			break
+		}
+		if _, ok := batchMessageIdToMessageBody[batchMessageId]; ok {
+			createErrorResponse(w, req, "BatchEntryIdsNotDistinct")
+			return
+		}
+
+		thisMessageBody := req.FormValue(thisMessageFormKey + ".Message")
+		thisMessageStructure := req.FormValue(thisMessageFormKey + ".MessageStructure")
+		thisMessageSubject := req.FormValue(thisMessageFormKey + ".Subject")
+
+		thisMessageAttributes := make(map[string]app.MessageAttributeValue)
+
+		for i := 1; true; i++ {
+			name := req.FormValue(fmt.Sprintf("%s.MessageAttributes.entry.%d.Name", thisMessageFormKey, i))
+			if name == "" {
+				break
+			}
+
+			dataType := req.FormValue(fmt.Sprintf("%s.MessageAttributes.entry.%d.Value.DataType", thisMessageFormKey, i))
+			if dataType == "" {
+				log.Warnf("DataType of %s.MessageAttribute %s is missing, MD5 checksum will most probably be wrong!\n", thisMessageFormKey, name)
+				continue
+			}
+
+			// StringListValue and BinaryListValue is currently not implemented
+			for _, valueKey := range [...]string{"StringValue", "BinaryValue"} {
```

Comment:
If someone provided both `StringValue` and `BinaryValue` would we overwrite `StringValue`?  If so, is that what we want?

Reply:
Reading the docs this needs refinement. I took this from the SNS:Publish (gosns.Publish) implementation (which may also need some rework I'm not going to touch it here).

Reply:
EDIT: Correction, the SNS docs are unclear on the behaviour here. I'll do what the localstack solution does...

Reply:
Sounds good, I like your pattern here better anyway - it's clearer about what we're doing.  We can dig deeper on how exactly it's supposed to work later.